docs/create_your_own_brick.rst

```diff
@@ -0,0 +1,364 @@
+:class:`.Application`, named ``apply``, which becomes an attribute of the
+brick.
+
+In the previous examples, variables were named with strings. If you need to
+name certain variables with other variables (such as ``self .fancy_name``),
+you should define them with the apply.property decorator:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     fancy_name = "salut_toi"
+    ...     @application
+    ...     def apply(self, input)
+    ...         ...
+    ...     @apply.property('inputs')
+    ...     def apply_inputs(self):
+    ...         return self.fancy_name
+
+You can also annotate specific variables by passing ``application_call`` as
+argument of your ``apply`` function, as shown in this example:
+
+    >>> class Foo(Brick): # doctest: +SKIP
+    ...     @application
+    ...     def apply(self, x, application_call):
+    ...         application_call.add_auxiliary_variable(x.mean())
+    ...         return x + 1
+
+You can retrieve these annotated variables as usual with the computational
+graph.
+
+
+Lazy initialization
+~~~~~~~~~~~~~~~~~~~
+
+.. doctest::
+   :hide:
+
+    >>> from blocks.bricks import Feedforward, Linear
+    >>> from blocks.bricks.base import lazy, application
+
+Instead of forcing the user to provide all the brick attributes as arguments
+to the :meth:`.Brick.__init__` method, you could let him/her specify them
+later, after the creation of the brick. To enable this mecanism, called lazy
+initialization, you need to decorate the method :meth:`.Brick.__init__` with
+the :func:`.lazy` decorator:
+
+    >>> @lazy(allocation=['attr1', 'attr2']) # doctest: +SKIP
+    ... def __init__(self, attr1, attr1)
+    ...     ...
+
+This allows the user to specify attr1 and attr2 after the creation of the
+brick. For example, the following ``DoubleSequential`` brick is composed of two
+:class:`.Feedforward` bricks for which you do not need to specify the
+``input_dim`` of brick2 directly at its creation.
+
+    >>> class DoubleSequential(Feedforward):
+    ...     """
+    ...     Two sequential Feedforward bricks.
+    ...     """
+    ...     def __init__(self, brick1, brick2, **kwargs):
+    ...         super(Feedforward, self).__init__(**kwargs)
+    ...         self.brick1 = brick1
+    ...         self.brick2 = brick2
+    ...         self.input_dim = brick1.input_dim
+    ...         self.output_dim = brick2.output_dim
+    ...         self.children = [self.brick1, self.brick2]
+    ...
+    ...     def _push_allocation_config(self):
+    ...         self.brick2.input_dim = self.brick1.get_dim('input_')
+    ...
+    ...     @application
+    ...     def apply(self, x):
+    ...         return self.brick2.apply(self.brick1.apply(x))
+
+Note how ``get_dim`` is used to retrieve the ``input_dim`` of ``brick1``. You
+can now create a DoubleSeuential brick as follows.
+
+    >>> brick1 = Linear(input_dim=3, output_dim=2)
+    >>> brick2 = Linear(output_dim=4)
+    >>>
+    >>> seq = DoubleSequential(brick1, brick2)
```

Comment:
It worth adding initialization and demonstrate that the input dimension of brick2 was set:
```
seq.initialize()
brick2.input_dim
```